package-lock.json

```diff
@@ -20855,7 +20821,7 @@
 				},
 				"nopt": {
 					"version": "4.0.1",
-					"resolved": "https://registry.npmjs.org/nopt/-/nopt-4.0.1.tgz",
+					"resolved": false,
```

Comment:
Heya, these changes aren't expected and cause local changes when running `npm install` on the `master` branch. No fault intended, I just want to better understand how this might have happened, since it was my hope these issues would have been resolved with the changes introduced in #21306.

Can you verify for me which version of NPM you're running locally? You can check by `npm -v`. Can you elaborate more on what lockfile issues you were remarking about in https://github.com/WordPress/gutenberg/pull/21873#discussion_r416508564 ?

Reply:
Fixed in e110119.

Reply:
I'm curious about how Travis is giving a ✅ for these changes as well. Incorrect package lock changes have slipped in a few times now.

Reply:
At least on what I'd been debugging at https://github.com/WordPress/gutenberg/issues/16157#issuecomment-599589261 , I think the issues with Travis are a combination of:

- These are optional dependencies, and specifically in the Linux environment that Travis runs in, they are skipped.
- NPM doesn't (didn't?) do a great job of verifying and updating lockfile for optional dependencies which are skipped.

Reply:
@aduth thank you for quickly spotting and fixing the issue! It's pretty interesting - if you take a look at the commit history on this PR, you'll see I've had some trouble with getting the package-lock.json check to work. What finally did the trick was a fresh clone of the repo, ensuring npm `6.14.4`, and doing `npm install` on this fresh clone. As a result, we ended up with this `"resolved": false` line. 

See these two commits specifically:

* https://github.com/WordPress/gutenberg/pull/21873/commits/1a2c1d203a319cf8420aa3c9977092521e583786 - changed `resolved` to `""` and Travis check failed - 👍 
* https://github.com/WordPress/gutenberg/pull/21873/commits/a7d3436f8721cc2e7f1d26590f3154f443c2fe15 - changed `resolved` to `false` and Travis check succeeded - 👎 

I am not sure why:

1. My npm generated the wrong `package-lock.json`
2. Travis happily accepted that

If I had to guess, I would say that either the check is not watertight, OR that something was going on with npm API at that time.

Reply:
Hm, strange indeed. I still think the reason why Travis is passing, based on the prior comment at https://github.com/WordPress/gutenberg/issues/16157#issuecomment-599589261, is that `node-pre-gyp` is not installed in Linux. At least, it's a transitive dependency of `fsevents`, where `fsevents` is explicitly skipped:

>npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.9 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.9: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

https://travis-ci.com/github/WordPress/gutenberg/jobs/325042262

I would have hoped that running the latest NPM would be enough for these sorts of issues to be avoided. I have noticed a few other cases lately where there seems to be a difference if running `npm install` with an existing `node_modules` vs. a fresh clone (e.g. https://github.com/WordPress/gutenberg/pull/21994#issuecomment-623454590). It sounds similar to what you describe. I'm not sure how best we could address this, other than assuming it's a bug in NPM and hoping for an upstream fix.

I had contemplated at one point if it might be a good idea to run this specific Travis job in a macOS environment, since it is an option:

https://docs.travis-ci.com/user/reference/osx/

I'd be worried though that it could still leave some variability of accuracy across other OS.

Reply:
@aduth if all these cases are about `"resolved": false` in package-lock.json, maybe testing for that would be a good workaround? As in `cat package-lock.json | grep `"resolved": false`?

Reply:
@adamziel That's an interesting idea, for sure. The issue almost always manifests itself this way [1]. There is some prior art here with [processing `git diff`](https://github.com/WordPress/gutenberg/blob/master/bin/process-git-diff.js) to account for some variance [2]. We could do something similar to check for these sorts of values in `package-lock.json`.

---

_[1] In the past there's also been some inconsistencies in toggling between `http` and `https`, though I've seen less of those lately._

_[2] Actually not sure if this specific script is even needed anymore._

Reply:
> @aduth if all these cases are about `"resolved": false` in package-lock.json, maybe testing for that would be a good workaround? As in `cat package-lock.json | grep "resolved": false`?

Proposed at #22237.

I forgot you had suggested a simple approach using `grep`. It can work, though the approach in #22237 actually validates against the expected structure of a `package-lock.json`. Maybe overkill, but the work's done already 🤷